### PR TITLE
CI environments overhaul

### PR DESCRIPTION
**Pull request by crusaderky**
Propaedeutic CI work to #3222.

- py36 and py37 are now identical
- Many optional dependencies were missing in one test suite or another (see details below)
- Tests that require hypothesis now always run if hypothesis is installed
- py37-windows.yml requirements file has been rebuilt starting from py37.yml
- Sorted requirements files alphabetically for better maintainability
- Added black. This is not needed by CI, but I personally use these yaml files to deploy my dev environment and I would expect many more developers to do the same.
  Alternatively, we could go the other way around and remove flake8 from everywhere and mypy from py36 and py37-windows. IMHO the marginal speedup would not be worth the complication.

Added packages to py36.yml (net of changes in order):
+ black
+ hypothesis
+ nc-time-axis
+ numba
+ numbagg
+ pynio (https://github.com/pydata/xarray/issues/3154 seems to be now fixed upstream)
+ sparse

Added packages to py37.yml (net of changes in order):

+ black
+ cdms2
+ hypothesis
+ iris>=1.10
+ numba (previously implicitly installed from pip by numbagg; now installed from conda)
+ pynio

Added packages to py37-windows.yml (net of changes in order):

+ black
+ bottleneck
+ flake8
+ hypothesis
+ iris>=1.10
+ lxml
+ mypy==0.720
+ numba
+ numbagg
+ pseudonetcdf>=3.0.1
+ pydap
+ sparse

### CODE REVIEW

**Comment by crusaderky** [comment]
Differences between the requirements files after the changes:

```patch
--- py36.yml
+++ py37.yml
@@ -5 +5 @@
-  - python=3.6
+  - python=3.7

--- py37.yml
+++ py37-windows.yml
@@ -9 +9 @@
-  - cdms2
+  # - cdms2  # Not available on Windows
@@ -31 +31 @@
-  - pynio
+  # - pynio  # Not available on Windows
```

**Comment by rabernat** [comment]
For this PR, it could be worth considering the doc built challenges we are discussing over in https://github.com/pydata/xarray/pull/3199#issuecomment-534702846.

In that case, the doc build succeeded on CI but still failed on RTD. In theory, that shouldn't be possible.

File: setup.cfg [hunk 1]
```diff
@@ -3,7 +3,7 @@ universal = 1
 
 [tool:pytest]
 python_files=test_*.py
-testpaths=xarray/tests
+testpaths=xarray/tests properties
```
Comment (max-sixty) [inline setup.cfg]: IIUC, do we want pytest running hypothesis tests?

Comment (crusaderky) [reply setup.cfg]: Why not? Is there a risk that they'll randomly start failing?

Note the added line to the test script that disables the whole thing if the hypothesis package is not installed.

**Comment by crusaderky** [comment]
@rabernat I'll look into it but on first inspection it feels like scope creep to me and I think it should be dealt with separately. Note how I didn't touch ``doc/environment.yml``.

File: setup.cfg (hunk 1 above)
Comment (crusaderky) [reply setup.cfg]: Also the whole test suite runs in 900ms for me

**Comment by crusaderky** [comment]
Ready for review

File: setup.cfg (hunk 1 above)
Comment (max-sixty) [reply setup.cfg]: Um, OK, then I take that back. I thought hypothesis took much longer!

Comment (crusaderky) [reply setup.cfg]: I've added ``@pytest.mark.slow``; that should settle the argument I believe?

**Comment by shoyer** [comment]
This looks great to me!

@crusaderky when you're ready, feel free to merge

File: setup.cfg (hunk 1 above)
Comment (max-sixty) [reply setup.cfg]: Great!

**Comment by crusaderky** [comment]
Thanks all!

**Comment by shoyer** [comment]
thank you, for tackling this!

**Pull request merged**